packages/wasm-utils/src/proof/mod.rs

```diff
@@ -1,511 +1,348 @@
+	pub fn set_relayer(&mut self, relayer: JsString) {
+		let r: String = relayer.into();
+		self.relayer = Some(hex::decode(r).unwrap());
 	}
 
-	pub fn set_note(&mut self, note: Note) {
-		self.note = Some(note);
+	#[wasm_bindgen(js_name = setLeaves)]
+	pub fn set_leaves(&mut self, leaves: Leaves) {
+		let ls: Vec<_> = Array::from(&leaves)
+			.to_vec()
+			.into_iter()
+			.map(|v| Uint8Array::new_with_byte_offset_and_length(&v, 0, 32))
+			.map(Uint8Arrayx32::try_from)
+			.collect::<Result<Vec<_>, _>>()
+			.unwrap()
+			.into_iter()
+			.map(|v| v.0)
+			.collect();
+		self.leaves = Some(ls);
 	}
 
-	pub fn set_fee(&mut self, fee: u32) {
-		self.fee = fee;
+	#[wasm_bindgen(js_name = setLeafIndex)]
+	pub fn set_leaf_index(&mut self, leaf_index: JsString) {
+		let leaf_index: String = leaf_index.into();
+		let leaf_index: u32 = leaf_index.as_str().parse().unwrap();
+		self.leaf_index = Some(leaf_index);
 	}
 
-	pub fn set_leaf_index(&mut self, leaf_index: u32) {
-		self.leaf_index = leaf_index;
+	#[wasm_bindgen(js_name = setFee)]
+	pub fn set_fee(&mut self, fee: JsString) {
+		let fee: String = fee.into();
+		let fee: u32 = fee.as_str().parse().unwrap();
+		self.fee = Some(fee);
 	}
 
-	pub fn set_refund(&mut self, refund: u32) {
-		self.refund = refund;
+	#[wasm_bindgen(js_name = setRefund)]
+	pub fn set_refund(&mut self, refund: JsString) {
+		let refund: String = refund.into();
+		let refund: u32 = refund.as_str().parse().unwrap();
+		self.refund = Some(refund);
 	}
 
-	pub fn set_proving_key(&mut self, pk_bytes: &[u8]) {
-		self.pk = pk_bytes.to_vec()
+	#[wasm_bindgen(js_name = setPk)]
+	pub fn set_pk(&mut self, pk: JsString) {
+		let p: String = pk.into();
+		self.pk = Some(hex::decode(p).unwrap());
 	}
 
-	pub fn build(&self) -> ZKProof {
-		ZKProof::new(self)
+	#[wasm_bindgen]
+	pub fn build(self) -> ProofInput {
+		let pk = self.pk.unwrap();
+		let recipient = self.recipient.unwrap();
+		let relayer = self.relayer.unwrap();
+
+		let leaf_index = self.leaf_index.unwrap();
+		let leaves = self.leaves.unwrap();
+
+		let fee = self.fee.unwrap();
+		let refund = self.refund.unwrap();
+
+		ProofInput {
+			pk,
+			relayer: truncate_and_pad(&relayer),
+			recipient: truncate_and_pad(&recipient),
+			refund,
+			fee,
+			leaf_index,
+			leaves,
+		}
 	}
 }
 
+#[wasm_bindgen]
+pub fn generate_proof_js(js_note: JsNote, proof_input: ProofInput) -> Result<Proof, JsValue> {
+	let note = js_note.note;
+	let width = note.width;
+	let exponentiation = note.exponentiation;
+	let backend = note.backend;
+	let curve = note.curve;
+	let note_secrets = note.secret;
+	let secrets = note_secrets[..32].to_vec();
+	let nullifier = note_secrets[32..64].to_vec();
+	let leaves: Vec<_> = proof_input.leaves.into_iter().map(|i| i.to_vec()).collect();
+	let leaf_index = proof_input.leaf_index as u64;
+	let recipient_raw = proof_input.recipient;
+	let relayer_raw = proof_input.relayer;
+
+	let fee = proof_input.fee as u128;
+	let refund = proof_input.refund as u128;
```

Comment:
Lets change leaf, fee, and refund types to match these. So leaf: `u64`, fee: `u128`, refund: `u128`